csrc/multidevice/lower_communication.cpp

```diff
@@ -397,10 +391,19 @@ void lowerToReduceScatter(
   params.redOp = getC10dReduceOpType(op_type);
   params.team = mesh.vector();
   params.dst_bufs = {output_tensor};
+  // TODO: Clean up once allocation domain for sharded tensors is
+  // better supported.
+  auto red_axis = output_tv->getReductionAxis().value();
+  auto shard_axis = getShardedAxis(output_tv);
+  if (red_axis <= shard_axis) {
+    shard_axis++;
+  }
   for (auto i : c10::irange(mesh.vector().size())) {
-    auto sliced_buf = input_tensor.index(
-        {at::indexing::Slice(0, 1), static_cast<int>(i), "..."});
-    params.src_bufs.push_back(sliced_buf);
+    std::vector<at::indexing::TensorIndex> indices(
+        input_tensor.dim(), at::indexing::Slice());
+    indices[shard_axis] = at::indexing::Slice(i, i + 1);
+    auto slice = input_tensor.index(indices).squeeze(red_axis);
```

Comment:
I think it can be simplified using the same trick as https://github.com/NVIDIA/Fuser/pull/1946#discussion_r1532569253